![image](competi.png)

#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

In [26]:
import numpy as np
import pandas as pd
from sklearn import datasets
import seaborn as sns
from sklearn.feature_selection import RFE
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Lasso
from sklearn.linear_model import ElasticNet
from sklearn.tree import DecisionTreeRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error
from sklearn.ensemble import RandomForestRegressor
from sklearn.datasets import make_regression
from sklearn.metrics import r2_score,mean_squared_error
import urllib.request
from PIL import Image
from xgboost import XGBClassifier
from sklearn.metrics import mean_absolute_error

In [3]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
df = pd.read_csv("diamonds_train.csv", index_col=0)

In [4]:
df.head()

carat        cut color clarity  depth  table     x     y     z  price
0   1.21      Ideal     H     VS2   63.0   57.0  6.73  6.70  4.23   6134
1   0.28  Very Good     D    VVS2   64.0   56.0  4.14  4.17  2.66    532
2   0.42    Premium     F     VS1   61.2   58.0  4.86  4.82  2.96   1103
3   0.26      Ideal     H      IF   61.1   57.0  4.16  4.12  2.53    600
4   1.10       Good     G     SI1   63.4   57.0  6.52  6.55  4.14   4997

In [13]:
df = pd.get_dummies(df)
features

carat  depth  table     x     y     z  price  cut_Fair  cut_Good  \
0       1.21   63.0   57.0  6.73  6.70  4.23   6134         0         0   
1       0.28   64.0   56.0  4.14  4.17  2.66    532         0         0   
2       0.42   61.2   58.0  4.86  4.82  2.96   1103         0         0   
3       0.26   61.1   57.0  4.16  4.12  2.53    600         0         0   
4       1.10   63.4   57.0  6.52  6.55  4.14   4997         0         1   
...      ...    ...    ...   ...   ...   ...    ...       ...       ...   
40340   1.55   61.3   61.0  7.46  7.39  4.55  11708         0         0   
40341   0.36   60.6   56.0  4.58  4.63  2.79    619         0         0   
40342   0.57   62.2   55.0  5.33  5.34  3.32   1267         0         0   
40343   1.01   59.6   62.0  6.47  6.56  3.88   9965         0         0   
40344   0.54   60.4   57.0  5.33  5.27  3.20   1340         0         0   

       cut_Ideal  ...  color_I  color_J  clarity_I1  clarity_IF  clarity_SI1  \
0              1  ...        0        0           0           0            0   
1              0  ...        0        0           0           0            0   
2              0  ...        0        0           0           0            0   
3              1  ...        0        0           0           1            0   
4              0  ...        0        0           0           0            1   
...          ...  ...      ...      ...         ...         ...          ...   
40340          0  ...        0        0           0           0            0   
40341          1  ...        0        0           0           0            1   
40342          0  ...        1        0           0           0            0   
40343          0  ...        0        0           0           1            0   
40344          1  ...        0        0           0           0            0   

       clarity_SI2  clarity_VS1  clarity_VS2  clarity_VVS1  clarity_VVS2  
0                0            0            1             0             0  
1                0            0            0             0             1  
2                0            1            0             0             0  
3                0            0            0             0             0  
4                0            0            0             0             0  
...            ...          ...          ...           ...           ...  
40340            0            0            1             0             0  
40341            0            0            0             0             0  
40342            0            0            1             0             0  
40343            0            0            0             0             0  
40344            1            0            0             0             0  

[40345 rows x 27 columns]

In [48]:
y = np.array(features['price'])

In [14]:
#feature_list = list(features.columns)
features.columns

Index(['carat', 'depth', 'table', 'x', 'y', 'z', 'price', 'cut_Fair',
       'cut_Good', 'cut_Ideal', 'cut_Premium', 'cut_Very Good', 'color_D',
       'color_E', 'color_F', 'color_G', 'color_H', 'color_I', 'color_J',
       'clarity_I1', 'clarity_IF', 'clarity_SI1', 'clarity_SI2', 'clarity_VS1',
       'clarity_VS2', 'clarity_VVS1', 'clarity_VVS2'],
      dtype='object')

In [81]:
X = np.array(features[['carat','depth', 'table', 'x', 'y', 'z', 'cut_Fair',
       'cut_Good', 'cut_Ideal', 'cut_Premium', 'cut_Very Good', 'color_D',
       'color_E', 'color_F', 'color_G', 'color_H', 'color_I', 'color_J',
       'clarity_I1', 'clarity_IF', 'clarity_SI1', 'clarity_SI2', 'clarity_VS1',
       'clarity_VS2', 'clarity_VVS1', 'clarity_VVS2']])

In [82]:
train_labels, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

In [83]:
rf = RandomForestRegressor(n_estimators=500, random_state=42)

In [84]:
rf.fit(train_labels, y_train)

RandomForestRegressor(n_estimators=500, random_state=42)

In [85]:
predictions = rf.predict(X_test)
# Calculate the absolute errors MAE
errors = abs(predictions - y_test)
# Print out the mean absolute error (mae)
print('Mean Absolute Error:', round(np.mean(errors), 2))

Mean Absolute Error: 282.35


In [86]:
np.sqrt(mean_squared_error(y_test, predictions))

577.2729548734748

In [87]:
mean_absolute_error(y_test, predictions)

282.3450864829536

**Importante:**

   - Si quitas columnas o creas columnas nuevas a partir de otras, o cualquier modificación column-wise tendrás que aplicarlo al dataset de `titanic_test.csv` de cara a hacer la predicción.
   - Si por lo contrario, decides por ejemplo, quitar los outliers o hacer un `dropna()`, o cualquier modificación row-wise eso NO PODRÁS (ni debes) aplicarlo al dataset de `titanic_test.csv` de cara a hacer la predicción. ¿Por qué? Porque si el conjunto de test tiene 50 observaciones (filas) la predicción se espera que tenga 50 filas.

#### ¿Y si lo que hago es estandarizar los datos o hacer un encoding, también se lo tengo que hacer al test antes de predecir?

In [5]:
# ¿qué opináis? 

-----------------------------------------------------------------------------------------------------------------

### Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

#### ¿Y si después del análisis exploratorio de mis datos llego a la conclusión de que puedo predecir el precio solo con las columnas `x`, `y` y `z`, también aplica al test?

In [8]:
# ¿qué opináis?

### 2. Dividir X_train, X_test, y_train, y_test

### 3. Asignar el modelo (vacío) a una variable
Aquí meteríais los parámetros. 

**Consejo**: Usa GridSearch y vuélvete loca o loco probando modelos e hiperparámetros.

Siempre hay tiempo para una historia:
https://catboost.ai/docs/concepts/python-reference_catboostregressor.html

### 4. Entrenar el modelo con X_train, y_train

### 5. Predecir con el modelo ya entrenado con X_test

In [54]:
predictions = rf.predict(X_test)
print(predictions)

[5817.363 3523.227  599.325 ... 4807.213 2237.047 4104.801]


### 6. Sacar métricas, valorar el modelo 

Recuerdo que en la competición se va a evaluar con la métrica de RMSE.

In [55]:
np.sqrt(mean_squared_error(y_test, predictions))

570.1987021681476

-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

_Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `diamonds_test.csv` nos arriesgamos a que digamos, no salga lo esperado._

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `diamonds_train.csv` al completo.

**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**

Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `diamonds_test.csv` para predecir.

**¿De dónde saco `diamonds_test.csv`?**

In [41]:
X_pred = pd.read_csv("diamonds_test.csv")
X_pred.head()

Unnamed: 0  carat        cut color clarity  depth  table     x     y     z
0           0   0.30      Ideal     H     SI2   60.0   56.0  4.41  4.43  2.65
1           1   0.34      Ideal     D      IF   62.1   57.0  4.52  4.46  2.79
2           2   1.57  Very Good     I     VS2   60.3   58.0  7.58  7.55  4.56
3           3   0.31      Ideal     H     VS2   61.8   57.0  4.32  4.36  2.68
4           4   1.51       Good     I    VVS1   64.0   60.0  7.26  7.21  4.63

In [42]:
X_pred.shape

(13449, 10)

In [44]:
X_pred = pd.get_dummies(X_pred)
X_pred

Unnamed: 0  carat  depth  table     x     y     z  cut_Fair  cut_Good  \
0               0   0.30   60.0   56.0  4.41  4.43  2.65         0         0   
1               1   0.34   62.1   57.0  4.52  4.46  2.79         0         0   
2               2   1.57   60.3   58.0  7.58  7.55  4.56         0         0   
3               3   0.31   61.8   57.0  4.32  4.36  2.68         0         0   
4               4   1.51   64.0   60.0  7.26  7.21  4.63         0         1   
...           ...    ...    ...    ...   ...   ...   ...       ...       ...   
13444       13444   1.05   60.3   55.0  6.55  6.68  3.99         0         0   
13445       13445   0.27   61.9   56.0  4.13  4.16  2.56         0         0   
13446       13446   1.50   61.8   60.0  7.30  7.37  4.53         0         0   
13447       13447   1.70   60.2   61.0  7.79  7.68  4.66         0         0   
13448       13448   0.30   61.2   57.0  4.29  4.34  2.64         0         0   

       cut_Ideal  ...  color_I  color_J  clarity_I1  clarity_IF  clarity_SI1  \
0              1  ...        0        0           0           0            0   
1              1  ...        0        0           0           1            0   
2              0  ...        1        0           0           0            0   
3              1  ...        0        0           0           0            0   
4              0  ...        1        0           0           0            0   
...          ...  ...      ...      ...         ...         ...          ...   
13444          1  ...        0        0           0           0            0   
13445          0  ...        0        0           0           0            0   
13446          0  ...        0        0           0           0            0   
13447          0  ...        0        0           0           0            0   
13448          1  ...        0        0           0           0            0   

       clarity_SI2  clarity_VS1  clarity_VS2  clarity_VVS1  clarity_VVS2  
0                1            0            0             0             0  
1                0            0            0             0             0  
2                0            0            1             0             0  
3                0            0            1             0             0  
4                0            0            0             1             0  
...            ...          ...          ...           ...           ...  
13444            1            0            0             0             0  
13445            0            1            0             0             0  
13446            0            0            1             0             0  
13447            1            0            0             0             0  
13448            0            0            1             0             0  

[13449 rows x 27 columns]

In [56]:
X_pred = np.array(X_pred[['carat','table', 'x', 'y', 'z', 'cut_Fair',
       'cut_Good', 'cut_Ideal', 'cut_Premium', 'cut_Very Good', 'color_D',
       'color_E', 'color_F', 'color_G', 'color_H', 'color_I', 'color_J',
       'clarity_I1', 'clarity_IF', 'clarity_SI1', 'clarity_SI2', 'clarity_VS1',
       'clarity_VS2', 'clarity_VVS1', 'clarity_VVS2']])

**¿Por qué da error?** 

In [57]:
predictions_submit = rf.predict(X_pred)
predictions_submit

array([  447.607,  2049.04 ,  9748.422, ..., 14785.937,  9901.024,
         767.578])

## IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 13449 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `diamonds_test.csv`?**

In [ ]:
# ¿Qué opináis?

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

**¿Qué más habrá que quitar o hacer?**

In [ ]:
# ¿Qué opináis?

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

**Truqui**: Carga X que debe estar actualmente actualizada a cómo la usaste para entrenar a `model` y haz que `to_pred` sea igual. ¡Sin quitar filas!

In [22]:
X_pred = np.array(X_pred[['x','y','z']])

Porque:
    - SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 3 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [ ]:
# respuestas

In [58]:
predictions_submit = rf.predict(X_pred)
predictions_submit

array([  447.607,  2049.04 ,  9748.422, ..., 14785.937,  9901.024,
         767.578])

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

**¿De dónde saco `diamonds_test.csv`?**

In [59]:
sample = pd.read_csv("sample_submission.csv")

In [60]:
sample.head()

id  price
0   0  12132
1   1  11786
2   2  14684
3   3  15425
4   4   6724

In [61]:
sample.shape

(13449, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [62]:
submission = pd.DataFrame({"id": range(len(predictions_submit)), "price": predictions_submit})

In [63]:
submission.head()

id     price
0   0   447.607
1   1  2049.040
2   2  9748.422
3   3   488.849
4   4  8773.017

In [64]:
submission.shape

(13449, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [65]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.id.all() == sample.id.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [45]:
help(chequeator)

Help on function chequeator in module __main__:

chequeator(df_to_submit)
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.



In [66]:
chequeator(submission)

You're ready to submit!


URLError: <urlopen error [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond>